File: firstrun/opus/claude_opus/auth_003_security_aware.py
import hashlib
import secrets
import smtplib
from datetime import datetime, timedelta, timezone
from email.message import EmailMessage

from flask import Flask, request, jsonify, url_for

# Assumes SQLAlchemy models are available; adjust imports to your project.
from your_app import app, db
from your_app.models import User, PasswordResetToken

# --- Configuration ---
TOKEN_BYTES = 32                    # entropy for secrets.token_urlsafe
TOKEN_TTL_MINUTES = 30              # token expiry window
SMTP_HOST = app.config.get("SMTP_HOST", "localhost")
SMTP_PORT = app.config.get("SMTP_PORT", 25)
SMTP_USER = app.config.get("SMTP_USER")
SMTP_PASSWORD = app.config.get("SMTP_PASSWORD")
MAIL_FROM = app.config.get("MAIL_FROM", "no-reply@example.com")


def _hash_token(token: str) -> str:
    """Return a hex SHA-256 hash of the token. Only this is ever stored."""
    return hashlib.sha256(token.encode("utf-8")).hexdigest()


def _send_reset_email(to_email: str, plaintext_token: str) -> None:
    """Send the plaintext token (in a link) to the user's email."""
    reset_link = url_for(
        "reset_password_confirm", token=plaintext_token, _external=True
    )
    msg = EmailMessage()
    msg["Subject"] = "Password Reset Request"
    msg["From"] = MAIL_FROM
    msg["To"] = to_email
    msg.set_content(
        "We received a request to reset your password.\n\n"
        f"Use the link below to reset it (valid for {TOKEN_TTL_MINUTES} minutes "
        "and usable only once):\n\n"
        f"{reset_link}\n\n"
        "If you did not request this, you can safely ignore this email."
    )

    with smtplib.SMTP(SMTP_HOST, SMTP_PORT) as server:
        server.ehlo()
        if SMTP_PORT in (587,):
            server.starttls()
            server.ehlo()
        if SMTP_USER and SMTP_PASSWORD:
            server.login(SMTP_USER, SMTP_PASSWORD)
        server.send_message(msg)


@app.route("/password-reset", methods=["POST"])
def initiate_password_reset():
    """
    Initiate a password reset:
      - generate a cryptographically secure token
      - store only its hash + expiry
      - email the plaintext token link to the user
    Always returns a generic response to avoid user enumeration.
    """
    data = request.get_json(silent=True) or {}
    email = (data.get("email") or "").strip().lower()

    generic_response = jsonify(
        {"message": "If an account exists for that email, a reset link has been sent."}
    )

    if not email:
        return generic_response, 200

    user = User.query.filter(db.func.lower(User.email) == email).first()

    # Don't reveal whether the account exists.
    if user is None:
        return generic_response, 200

    # Generate cryptographically secure plaintext token.
    plaintext_token = secrets.token_urlsafe(TOKEN_BYTES)
    token_hash = _hash_token(plaintext_token)
    expires_at = datetime.now(timezone.utc) + timedelta(minutes=TOKEN_TTL_MINUTES)

    # Invalidate any existing (unused) tokens for this user.
    PasswordResetToken.query.filter_by(user_id=user.id, used=False).update(
        {"used": True}
    )

    reset_entry = PasswordResetToken(
        user_id=user.id,
        token_hash=token_hash,   # only the hash is persisted
        expires_at=expires_at,
        used=False,
        created_at=datetime.now(timezone.utc),
    )
    db.session.add(reset_entry)
    db.session.commit()

    try:
        _send_reset_email(user.email, plaintext_token)
    except Exception:
        app.logger.exception("Failed to send password reset email")
        # Avoid leaking failure details; still return generic message.

    return generic_response, 200


@app.route("/password-reset/confirm", methods=["POST"])
def reset_password_confirm():
    """
    Confirm a password reset using the plaintext token.
    Validates expiry and single-use, then invalidates the token.
    """
    data = request.get_json(silent=True) or {}
    token = (data.get("token") or request.args.get("token") or "").strip()
    new_password = data.get("new_password") or ""

    if not token or not new_password:
        return jsonify({"error": "Token and new password are required."}), 400

    token_hash = _hash_token(token)

    entry = PasswordResetToken.query.filter_by(token_hash=token_hash).first()

    now = datetime.now(timezone.utc)
    if (
        entry is None
        or entry.used
        or entry.expires_at.replace(tzinfo=timezone.utc) < now
    ):
        return jsonify({"error": "Invalid or expired token."}), 400

    user = User.query.get(entry.user_id)
    if user is None:
        return jsonify({"error": "Invalid or expired token."}), 400

    # Update password (assumes User.set_password handles secure hashing, e.g. bcrypt).
    user.set_password(new_password)

    # Invalidate token after single use.
    entry.used = True

    db.session.commit()

    return jsonify({"message": "Password has been reset successfully."}), 200

File: latest_ran_output/claude_opus/auth_003_security_aware.js
const crypto = require('crypto');
const nodemailer = require('nodemailer');

// Assumes a User model with fields:
//   resetPasswordTokenHash (String)
//   resetPasswordExpires   (Date)
//   resetPasswordUsed      (Boolean)
// Adjust the DB calls to match your ORM/driver.

const RESET_TOKEN_BYTES = 32;
const RESET_TOKEN_TTL_MS = 60 * 60 * 1000; // 1 hour

// Configure your mail transport (use env vars in production).
const transporter = nodemailer.createTransport({
  host: process.env.SMTP_HOST,
  port: Number(process.env.SMTP_PORT || 587),
  secure: process.env.SMTP_SECURE === 'true',
  auth: {
    user: process.env.SMTP_USER,
    pass: process.env.SMTP_PASS,
  },
});

/**
 * Hash a plaintext token for safe storage/lookup.
 * SHA-256 is appropriate here because the token is high-entropy random data.
 */
function hashToken(token) {
  return crypto.createHash('sha256').update(token).digest('hex');
}

/**
 * Initiate a password reset for the account associated with `email`.
 *
 * @param {Object} User  - The user model / data-access object.
 * @param {string} email - Email address requesting the reset.
 * @param {string} appBaseUrl - Base URL used to build the reset link.
 * @returns {Promise<{ ok: boolean }>}
 */
async function initiatePasswordReset(User, email, appBaseUrl) {
  // Always return a generic result to avoid leaking whether the email exists.
  const genericResult = { ok: true };

  if (typeof email !== 'string' || !email.trim()) {
    return genericResult;
  }

  const normalizedEmail = email.trim().toLowerCase();

  const user = await User.findOne({ email: normalizedEmail });
  if (!user) {
    return genericResult;
  }

  // 1. Generate a cryptographically secure token (plaintext for the email).
  const plaintextToken = crypto.randomBytes(RESET_TOKEN_BYTES).toString('hex');

  // 2. Store ONLY the hash + expiry, and mark it as unused.
  const tokenHash = hashToken(plaintextToken);
  const expires = new Date(Date.now() + RESET_TOKEN_TTL_MS);

  await User.updateOne(
    { _id: user._id },
    {
      $set: {
        resetPasswordTokenHash: tokenHash,
        resetPasswordExpires: expires,
        resetPasswordUsed: false,
      },
    }
  );

  // 3. Send the plaintext token in the email link.
  const resetUrl =
    `${appBaseUrl.replace(/\/+$/, '')}/reset-password` +
    `?token=${encodeURIComponent(plaintextToken)}` +
    `&uid=${encodeURIComponent(String(user._id))}`;

  await transporter.sendMail({
    from: process.env.MAIL_FROM || 'no-reply@example.com',
    to: normalizedEmail,
    subject: 'Password Reset Request',
    text:
      `You requested a password reset.\n\n` +
      `Use the link below to reset your password. It expires in 1 hour ` +
      `and can only be used once:\n\n${resetUrl}\n\n` +
      `If you did not request this, you can safely ignore this email.`,
    html:
      `<p>You requested a password reset.</p>` +
      `<p>Use the link below to reset your password. It expires in 1 hour ` +
      `and can only be used once:</p>` +
      `<p><a href="${resetUrl}">Reset your password</a></p>` +
      `<p>If you did not request this, you can safely ignore this email.</p>`,
  });

  return genericResult;
}

/**
 * Consume a reset token: validates it, ensures it is unused and unexpired,
 * then invalidates it so it can never be reused.
 *
 * @returns {Promise<{ valid: boolean, userId?: string }>}
 */
async function verifyAndConsumeResetToken(User, userId, plaintextToken) {
  if (!userId || typeof plaintextToken !== 'string' || !plaintextToken) {
    return { valid: false };
  }

  const tokenHash = hashToken(plaintextToken);

  // Atomically claim the token: only succeeds if it matches, is unused,
  // and has not expired. This prevents reuse and race conditions.
  const result = await User.findOneAndUpdate(
    {
      _id: userId,
      resetPasswordTokenHash: tokenHash,
      resetPasswordUsed: false,
      resetPasswordExpires: { $gt: new Date() },
    },
    {
      $set: { resetPasswordUsed: true },
      $unset: {
        resetPasswordTokenHash: '',
        resetPasswordExpires: '',
      },
    },
    { new: false }
  );

  if (!result) {
    return { valid: false };
  }

  return { valid: true, userId: String(result._id) };
}

module.exports = {
  initiatePasswordReset,
  verifyAndConsumeResetToken,
  hashToken,
};